bytecomp/matching.ml

```diff
@@ -1818,98 +1963,127 @@ let expand_stringswitch loc arg sw d = match d with
 (* Add handler, if shared *)
 let handle_shared () =
   let hs = ref (fun x -> x) in
-  let handle_shared act = match act with
-  | Switch.Single act -> act
-  | Switch.Shared act ->
-      let i,h = make_catch_delayed act in
-      let ohs = !hs in
-      hs := (fun act -> h (ohs act)) ;
-      make_exit i in
-  hs,handle_shared
-
+  let handle_shared act =
+    match act with
+    | Switch.Single act -> act
+    | Switch.Shared act ->
+        let i,h = make_catch_delayed act in
+        let ohs = !hs in
+        hs := (fun act -> h (ohs act));
+        make_exit act.block_loc i
+  in
+  hs, handle_shared
 
-let share_actions_tree sw d =
+let share_actions_tree sw ~(default : Lambda.block option) =
   let store = StoreExp.mk_store () in
-(* Default action is always shared *)
-  let d =
-    match d with
+  (* Default action is always shared *)
+  let default =
+    match default with
     | None -> None
-    | Some d -> Some (store.Switch.act_store_shared () d) in
-(* Store all other actions *)
+    | Some default -> Some (store.Switch.act_store_shared () default)
+  in
+  (* Store all other actions *)
   let sw =
-    List.map  (fun (cst,act) -> cst,store.Switch.act_store () act) sw in
-
-(* Retrieve all actions, including potential default *)
+    List.map (fun (cst,act) -> cst,store.Switch.act_store () act) sw
+  in
+  (* Retrieve all actions, including potential default *)
   let acts = store.Switch.act_get_shared () in
-
-(* Array of actual actions *)
+  (* Array of actual actions *)
   let hs,handle_shared = handle_shared () in
   let acts = Array.map handle_shared acts in
+  (* Reconstruct default and switch list *)
+  let default =
+    match default with
+    | None -> None
+    | Some default -> Some (acts.(default))
+  in
+  let sw = List.map (fun (cst, j) -> cst, acts.(j)) sw in
+  !hs,sw,default
+
+let first_location_in rem =
+  match rem with
+  | [] -> Location.none
+  | ((loc, _), _) :: _ -> loc
```

Comment:
There are four occurrences of `[] -> Location.none` in this file, all introduced by this patch. Maybe a generic function would be nice?

```ocaml
let first_location_with get_loc = function
  | [] -> Location.none
  | x :: _ -> get_loc x

let first_location_in rem = first_location_with fst rem
```

(and usages in `add_omega_column`, `record_matching_line` and `for_trywith`)

(These uses of `Location.none` look small code smells -- there is no good reason to actually try to build something with `Location.none` in this module -- but using them for now is the pragmatic thing to do for simplicity. Hopefully someday we will refactor the code and get rid of them.)

Reply:
I think I'd like to try to get rid of these `Location.none`s, but probably need help from @trefis to do so.